File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/Arrays/arrays.py
from typing import TypeVar, List, Set, Any, Optional
from collections import deque

T = TypeVar('T')

class ArrayOperations:
    @staticmethod
    def insert(array: List[T], element: T, position: int) -> List[T]:
        if position < 0 or position > len(array):
            raise ValueError("Invalid position")
        return array[:position] + [element] + array[position:]

    @staticmethod
    def delete(array: List[T], position: int) -> List[T]:
        if position < 0 or position >= len(array):
            raise ValueError("Invalid position")
        return array[:position] + array[position + 1:]

    @staticmethod
    def search(array: List[T], element: T) -> int:
        try:
            return array.index(element)
        except ValueError:
            return -1

    @staticmethod
    def update(array: List[T], element: T, position: int) -> None:
        if position < 0 or position >= len(array):
            raise ValueError("Invalid position")
        array[position] = element

    @staticmethod
    def reverse(array: List[T]) -> None:
        left, right = 0, len(array) - 1
        while left < right:
            array[left], array[right] = array[right], array[left]
            left += 1
            right -= 1

    @staticmethod
    def find_max(array: List[Any]) -> Optional[Any]:
        if not array:
            raise ValueError("Array is empty")
        return max(array)

    @staticmethod
    def find_min(array: List[Any]) -> Optional[Any]:
        if not array:
            raise ValueError("Array is empty")
        return min(array)

    @staticmethod
    def rotate(array: List[T], positions: int) -> None:
        if not array:
            return
        positions = positions % len(array)
        deque(array).rotate(-positions)

    @staticmethod
    def find_duplicates(array: List[T]) -> List[T]:
        seen: Set[T] = set()
        duplicates: Set[T] = set()
        for item in array:
            if item in seen:
                duplicates.add(item)
            seen.add(item)
        return list(duplicates) 

File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/Arrays/Arrays.cs
public class ArrayOperations
{
    // Basic Operations
    public static void Insert<T>(ref T[] array, T element, int position)
    {
        if (position < 0 || position > array.Length)
            throw new ArgumentOutOfRangeException("position");

        T[] newArray = new T[array.Length + 1];
        
        // Copy elements before position
        Array.Copy(array, 0, newArray, 0, position);
        
        // Insert new element
        newArray[position] = element;
        
        // Copy elements after position
        Array.Copy(array, position, newArray, position + 1, array.Length - position);
        
        array = newArray;
    }

    public static void Delete<T>(ref T[] array, int position)
    {
        if (position < 0 || position >= array.Length)
            throw new ArgumentOutOfRangeException("position");

        T[] newArray = new T[array.Length - 1];
        
        // Copy elements before position
        Array.Copy(array, 0, newArray, 0, position);
        
        // Copy elements after position
        Array.Copy(array, position + 1, newArray, position, array.Length - position - 1);
        
        array = newArray;
    }

    public static int Search<T>(T[] array, T element)
    {
        for (int i = 0; i < array.Length; i++)
        {
            if (array[i]?.Equals(element) == true)
                return i;
        }
        return -1;
    }

    public static void Update<T>(T[] array, T element, int position)
    {
        if (position < 0 || position >= array.Length)
            throw new ArgumentOutOfRangeException("position");

        array[position] = element;
    }

    public static void Reverse<T>(T[] array)
    {
        int left = 0;
        int right = array.Length - 1;

        while (left < right)
        {
            T temp = array[left];
            array[left] = array[right];
            array[right] = temp;
            left++;
            right--;
        }
    }

    public static T FindMax<T>(T[] array) where T : IComparable<T>
    {
        if (array.Length == 0)
            throw new ArgumentException("Array is empty");

        T max = array[0];
        for (int i = 1; i < array.Length; i++)
        {
            if (array[i].CompareTo(max) > 0)
                max = array[i];
        }
        return max;
    }

    public static T FindMin<T>(T[] array) where T : IComparable<T>
    {
        if (array.Length == 0)
            throw new ArgumentException("Array is empty");

        T min = array[0];
        for (int i = 1; i < array.Length; i++)
        {
            if (array[i].CompareTo(min) < 0)
                min = array[i];
        }
        return min;
    }

    // Common Array Algorithms
    public static void Rotate<T>(T[] array, int positions)
    {
        int n = array.Length;
        positions = positions % n;
        
        // Create temporary array to store rotated elements
        T[] temp = new T[positions];
        
        // Store first 'positions' elements in temp array
        Array.Copy(array, 0, temp, 0, positions);
        
        // Shift remaining elements to the left
        Array.Copy(array, positions, array, 0, n - positions);
        
        // Copy temp elements to end
        Array.Copy(temp, 0, array, n - positions, positions);
    }

    public static List<T> FindDuplicates<T>(T[] array)
    {
        List<T> duplicates = new List<T>();
        HashSet<T> seen = new HashSet<T>();

        foreach (T item in array)
        {
            if (!seen.Add(item) && !duplicates.Contains(item))
                duplicates.Add(item);
        }

        return duplicates;
    }
}
